File: src/decoders/discussion-message.ts
import { DiscussionMessage, EntityKind, SessionHandle } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { DiscussionMessageRecipient } from "~/models/discussion-message-recipient";
import { decodeAttachment } from "./attachment";

const hintToEntity = (hint: string): EntityKind => {
  let type: EntityKind;

  switch (hint) {
    case "Professeur":
      type = EntityKind.Teacher;
      break;
    case "Personnel":
      type = EntityKind.Personal;
      break;
    default:
      type = EntityKind.Student;
  }

  return type;
};

export const decodeDiscussionMessage = (message: any, session: SessionHandle): DiscussionMessage => {
  let author: DiscussionMessageRecipient | undefined;

  if (message.public_gauche !== "Moi")
    author = { name: message.public_gauche, kind: hintToEntity(message.hint_gauche) };

  let receiver: DiscussionMessageRecipient | undefined;

  if (message.public_droite === "Moi")
    receiver = undefined;
  else if (typeof message.public_droite === "string")
    receiver = { name: message.public_droite, kind: hintToEntity(message.hint_droite) };

  return {
    id: message.N,
    content: message.estHTML ? message.contenu.V : message.contenu,
    creationDate: decodePronoteDate(message.date.V),
    author, receiver,
    partialVisibility: message.estUnAparte,
    amountOfRecipients: (message.nbPublic ?? 1) + 1, // `+1` because the author is also a recipient.
    files: message.listeDocumentsJoints?.V.map((attachment: any) => decodeAttachment(attachment, session)) ?? []
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/discussion-message.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.DiscussionMessage
import ink.literate.pawnote.models.DiscussionMessageRecipient
import ink.literate.pawnote.models.EntityKind
import ink.literate.pawnote.models.SessionHandle
import kotlinx.serialization.json.*

fun hintToEntity (hint: String) : EntityKind {
    val type: EntityKind = when (hint) {
        "Professeur" -> EntityKind.Teacher
        "Personnel" -> EntityKind.Personal
        else -> EntityKind.Student
    }

    return type
}

fun <T> decodeDiscussionMessage(message: JsonObject, session: SessionHandle, decoder: (message: JsonObject) -> T): DiscussionMessage<T> {
    var author: DiscussionMessageRecipient? = null

    if (message["public_gauche"]!!.jsonPrimitive.content != "Moi")
        author = DiscussionMessageRecipient(
            name = message["public_gauche"]!!.jsonPrimitive.content,
            kind = hintToEntity(message["hint_gauche"]!!.jsonPrimitive.content)
        )

    var receiver: DiscussionMessageRecipient? = null

    if (message["public_droite"]!!.jsonPrimitive.isString)
        receiver = DiscussionMessageRecipient(
            name = message["public_droite"]!!.jsonPrimitive.content,
            kind = hintToEntity(message["hint_droite"]!!.jsonPrimitive.content)
        )

    return DiscussionMessage(
        id = message["N"]!!.jsonPrimitive.content,
        content = if(message["estHTML"]?.jsonPrimitive?.boolean == true) message["contenu"]!!.jsonObject["V"]!!.jsonPrimitive.content else message["contenu"]!!.jsonPrimitive.content,
        creationDate = decodePronoteDate(message["date"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        author = author,
        receiver = receiver,
        partialVisibility = message["estUnAparte"]!!.jsonPrimitive.boolean,
        amountOfRecipients = (message["nbPublic"]?.jsonPrimitive?.int ?: 1) + 1,
        files = if (message["listeDocumentsJoints"] != null) message["listeDocumentsJoints"]!!.jsonObject["V"]!!.jsonArray.map { decodeAttachment(it.jsonObject, session.information) } else listOf(),
        data = decoder(message)
    )
}